src/coreclr/vm/loaderallocator.cpp

```diff
@@ -384,6 +384,7 @@ LoaderAllocator * LoaderAllocator::GCLoaderAllocators_RemoveAssemblies(AppDomain
     AppDomain::AssemblyIterator i;
     // Iterate through every loader allocator, marking as we go
     {
+        CrstHolder chLoaderAllocatorReferencesLock(pAppDomain->GetLoaderAllocatorReferencesLock());
         CrstHolder chAssemblyListLock(pAppDomain->GetAssemblyListLock());
```

Comment:
The second iteration below is going to acquire same locks now. We can acquire the locks just once for the whole duration of both iterations.

Reply:
Makes sense, I'll update the PR that way.